R/trace_tests.R

```diff
@@ -227,18 +272,10 @@ truncate_call <- function(call_obj, limit = 1e4) {
 #' test call stack has changed, in which case we are onto a new test.
 #'
 is_current_test_finished <- function() {
-  syscalls <- sys.calls()
-
   is.null(.current_test$src) ||
-  .current_test$last_frame > length(syscalls) ||
-  !identical(
-    .current_test$src,
-    getSrcref(syscalls[[.current_test$last_frame]]) %||% syscalls[[.current_test$last_frame]]
-  ) ||
-  !identical(
-    .current_test$trace,
-    syscalls[.current_test$frames]
-  )
+  .current_test$last_frame > sys.nframe() ||
+  !identical(.current_test$src_call, sys.call(which = .current_test$last_frame)) ||
+  !identical(.current_test$src_env, sys.frame(which = .current_test$last_frame))
 }
```

Comment:
The major focus of this refactor is to reduce `sys.calls()` calls. The last call and environment in the call stack before we enter package namespace code is used as a heuristic to see if the test has changed. This is a bit less precise than when it was comparing the full call stack before, but is much, much faster. 

Situations where the same call, from the same environment actually corresponds to two different tests is possible, but highly unlikely. In my opinion, not worth the performance overhead of unwinding the call stack with each call to `covr::count()`.